src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataPrimitiveSerializer.cs

```diff
@@ -155,19 +156,54 @@ internal static object ConvertPrimitiveValue(object value, IEdmPrimitiveTypeRefe
             }
 
             Type type = value.GetType();
-            if (primitiveType != null && primitiveType.IsDate() && TypeHelper.IsDateTime(type))
+
+            if (primitiveType != null)
             {
-                Date dt = (DateTime)value;
-                return dt;
+                if (primitiveType.IsDate() && TypeHelper.IsDateTime(type))
+                {
+                    Date dt = (DateTime)value;
+                    return dt;
+                }
+
+                if (primitiveType.IsTimeOfDay() && TypeHelper.IsTimeSpan(type))
+                {
+                    TimeOfDay tod = (TimeSpan)value;
+                    return tod;
+                }
+
+                if (primitiveType is EdmDecimalTypeReference decimalTypeReference && decimalTypeReference.Scale.HasValue && value is decimal decimalValue)
+                {
+                    return CalculateScaleOfDecimal(decimalValue, decimalTypeReference.Scale.Value);
+                }
             }
 
-            if (primitiveType != null && primitiveType.IsTimeOfDay() && TypeHelper.IsTimeSpan(type))
+            return ConvertUnsupportedPrimitives(value);
+        }
+
+        static decimal CalculateScaleOfDecimal(decimal value, int scale)
+        {
+            string leftSide = GetLeftSideFromPoint(value);
+            string rightSide = GetRightSideFromPoint(value);
+            var numberOfZeroToAddToTheEndOfDecimalValue = scale - rightSide.Length;
+            for (int i = 0; i < numberOfZeroToAddToTheEndOfDecimalValue; i++)
             {
-                TimeOfDay tod = (TimeSpan)value;
-                return tod;
+                rightSide += "0";
```

Comment:
Since you know the size of the string in advance (i.e. `numberOfZeroAtoAddToTheEndOfDecimalValue`), you could pre-allocate block of chars on the stack using `stackalloc` or rent an array from the array pool if the size exceeds a certain threshold. Then you can add the `'0'` chars to the block and append to `rightSide` to avoid allocating all those temporary strings when doing `rightSide += "0"` inside of a loop.

e.g.
```c#
chart[] rentedArray = null;
Span<char>rightSideBuffer;
if (scale <= 256) {
   buffer = stackalloc char[scale];
}
else {
   rentedArray = ArrayPool<chart>.Rent(scale);
   buffer = rentedArray;
}

for (int i = 0; i < scale && i < rightSideLength; i++) {
    rightSideBuffer[rightSideStartPos + i] = valueAsString[rightSideStartPos + i];
}

for (int i = rightSideLength; i < scale; i++) {
   rightSdeBuffer[rightSideLength + i] = '0';
}

string rightSide = new string(rightSideBuffer.Slice(0, slice));

if (rentedArray != null) {
   ArrayPool<char>.Return(rentedArray);
}
```
Not sure if that ties all the screws, but hopefully the general idea is clear. Maybe we can probably also avoid allocating `leftSide`. 

We want to avoid "avoidable" allocations in methods that could be executed in a hot path, in this method will be executed for every decimal property when Scale is provided.